### PR TITLE
Fix: Finder incorrectly identifies pages created. [ED-3708]

### PR DESCRIPTION
**Pull request by iNewLegend**
image

### CODE REVIEW

**Review by Nevoss — APPROVED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/common/modules/finder/categories/create.php [hunk 1]
```diff
@@ -39,10 +38,44 @@ public function get_title() {
...
+		foreach ( $registered_document_types as $document_name => $document_type ) {
+			/**
+			 * @var $class \Elementor\Core\Base\Document
+			 */
+			$document_class = $document_type;
+			$document_properties = $document_class::get_properties();
+			$document_cpt = empty( $document_properties['cpt'] ) ? [] : $document_properties['cpt'];
+
+			if ( empty( $document_cpt ) ) {
+				continue;
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php:57]: What do you mean?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/base/document.php [hunk 2]
```diff
@@ -100,6 +100,10 @@ protected static function get_editor_panel_categories() {
 		return Plugin::$instance->elements_manager->get_categories();
 	}
 
+	public static function get_type() {
+		return 'post';
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply core/base/document.php]: Handled.

File: core/common/modules/finder/categories/create.php [hunk 3]
```diff
@@ -47,73 +47,98 @@ public function get_category_items( array $options = [] ) {
...
+					return false;
 				}
+
+				return true;
+			};
+
+			// To Support backward compatibility avoid those.
+			if ( 'post' === $document_name || 'page' === $document_name ) {
+				continue;
 			}
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled.

File: core/base/document.php [hunk 4]
```diff
@@ -205,10 +205,10 @@ public static function get_class_full_name() {
 
 	public static function get_create_url() {
 		$properties = static::get_properties();
-		$cpt = $properties['cpt'][0];
-		$base_create_url = Plugin::$instance->documents->get_create_new_post_url( $cpt );
 
-		return add_query_arg( [ 'template_type' => static::get_type() ], $base_create_url );
+		$cpt = 'wp-post' === static::get_type() ? 'post' : $properties['cpt'][0];
```
[the remark this answers is not in the document]
Comment (iNewLegend) [reply core/base/document.php]: Handled.

File: core/common/modules/finder/categories/create.php [hunk 5]
```diff
@@ -39,33 +38,97 @@ public function get_title() {
 	 * @return array
 	 */
 	public function get_category_items( array $options = [] ) {
+		$result = [];
+
+		$registered_document_types = Plugin::$instance->documents->get_document_types();
 		$elementor_supported_post_types = get_post_types_by_support( 'elementor' );
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:44]:
```suggestion
		$elementor_supported_post_types = get_post_types_by_support( 'elementor' );
		// Work with the array keys.
		$elementor_supported_post_types = array_flip( elementor_supported_post_types );
```

File: core/common/modules/finder/categories/create.php [hunk 6]
```diff
@@ -39,33 +38,97 @@ public function get_title() {
...
+			$document_properties = $document_class::get_properties();
+
+			if ( empty( $document_properties['show_in_finder'] ) ) {
+				continue;
+			}
+
+			// To Support backward compatibility, avoid those.
+			if ( 'post' === $document_name || 'page' === $document_name || 'stack' === $document_class::get_type() ) {
+				continue;
+			}
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:62]:
```suggestion
			if ( ! empty( $document_properties['cpt'] ) {
				foreach(  $document_properties['cpt'] as $cpt ) {
						unset( $elementor_supported_post_types[ $cpt ]  );
```

File: core/common/modules/finder/categories/create.php [hunk 7]
```diff
@@ -39,33 +38,97 @@ public function get_title() {
...
+			$document_types[ $document_name ] = $document_class;
+		}
+
+		// Will be handled by new mechanism.
+		$ignore_list = [
+			'post',
+			'page',
+			'e-landing-page',
+			'elementor_library',
+		];
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:73]:
```suggestion
		// Will be handled by new mechanism.
		$ignore_list = [
			'elementor_library',
		];
```

File: core/common/modules/finder/categories/create.php [hunk 8]
```diff
@@ -39,33 +38,97 @@ public function get_title() {
...
+			if ( empty( $document_properties['show_in_finder'] ) ) {
+				continue;
+			}
+
+			// To Support backward compatibility, avoid those.
+			if ( 'post' === $document_name || 'page' === $document_name || 'stack' === $document_class::get_type() ) {
+				continue;
+			}
+
+			$document_types[ $document_name ] = $document_class;
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:64]:
```suggestion
			$url = $this->create_item_url_by_document_class( $document_class );
			... // add to the finder
```

File: core/common/modules/finder/categories/create.php [hunk 9]
```diff
@@ -39,33 +38,97 @@ public function get_title() {
...
+		);
+	}
+
+	private function get_create_new_template( $title, $url ) {
+		return [
+			/* translators: %s the title of the post type */
+			'title' => sprintf( __( 'Add New %s', 'elementor' ), $title ),
+			'icon' => 'plus-circle-o',
+			'url' => $url,
+			'keywords' => [ 'post', 'page', 'template', 'new', 'create' ],
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:131]: Add the $title as a keyword

File: modules/library/documents/not-supported.php [hunk 10]
```diff
@@ -59,6 +60,10 @@ public static function get_title() {
 		return __( 'Not Supported', 'elementor' );
 	}
 
+	public static function get_create_url() {
+		return admin_url( Source_Local::ADMIN_MENU_SLUG . '#add_new' );
+	}
+
```
Comment (matipojo) [inline modules/library/documents/not-supported.php:66]: Remove

**Review by matipojo — CHANGES_REQUESTED** [review]

File: core/common/modules/finder/categories/create.php (hunk 5 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled.

File: core/common/modules/finder/categories/create.php (hunk 6 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled.

File: core/common/modules/finder/categories/create.php (hunk 7 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled.

File: core/common/modules/finder/categories/create.php (hunk 8 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled.

File: core/common/modules/finder/categories/create.php (hunk 9 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled.

File: modules/library/documents/not-supported.php (hunk 10 above)
Comment (iNewLegend) [reply modules/library/documents/not-supported.php]: Handled.

File: core/common/modules/finder/categories/create.php [hunk 11]
```diff
@@ -39,33 +38,92 @@ public function get_title() {
...
 			}
 
-			if ( Source_Local::CPT === $post_type ) {
-				$url = admin_url( Source_Local::ADMIN_MENU_SLUG . '#add_new' );
-			} else {
-				$url = Plugin::$instance->documents->get_create_new_post_url( $post_type );
+			// To Support backward compatibility.
+			if ( empty( $document_properties['cpt'] ) ) {
+				continue;
+			}
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:56]: Why?

File: core/common/modules/finder/categories/create.php [hunk 12]
```diff
@@ -39,33 +38,92 @@ public function get_title() {
...
+
+			foreach ( $document_properties['cpt'] as $cpt ) {
+				unset( $elementor_supported_post_types[ $cpt ] );
+			}
+
+			if ( in_array( $document_name, [ 'post', 'page', 'code_snippet' ], true ) ) {
+				continue;
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:63]: Why?

File: core/common/modules/finder/categories/create.php [hunk 13]
```diff
@@ -39,33 +38,92 @@ public function get_title() {
...
+
+			if ( ! $url ) {
+				continue;
+			}
+
+			$result[ $document_name ] = $url;
 		}
 
-		return $items;
+		// Handled by new mechanism.
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:76]:
```suggestion
```

File: core/common/modules/finder/categories/create.php [hunk 14]
```diff
@@ -39,33 +38,92 @@ public function get_title() {
...
+		$ignore_list = [
+			'elementor_library',
+		];
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:79]:
```suggestion
		unset( $elementor_supported_post_types['elementor_library'] ),
```

+ Explain

File: core/common/modules/finder/categories/create.php [hunk 15]
```diff
@@ -39,33 +38,92 @@ public function get_title() {
...
+
+		// Old mechanism.
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:81]:
```suggestion
```

File: core/common/modules/finder/categories/create.php [hunk 16]
```diff
@@ -39,33 +38,92 @@ public function get_title() {
...
+		foreach ( $elementor_supported_post_types as $post_type => $val ) {
+			if ( in_array( $post_type, $ignore_list, true ) ) {
+				continue;
+			}
```
Comment (matipojo) [inline core/common/modules/finder/categories/create.php:85]:
```suggestion
```

**Review by matipojo — CHANGES_REQUESTED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Fix comment + conflict

File: core/common/modules/finder/categories/create.php (hunk 11 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled

File: core/common/modules/finder/categories/create.php (hunk 13 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled

File: core/common/modules/finder/categories/create.php (hunk 14 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled

File: core/common/modules/finder/categories/create.php (hunk 15 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled

File: core/common/modules/finder/categories/create.php (hunk 16 above)
Comment (iNewLegend) [reply core/common/modules/finder/categories/create.php]: Handled

**Review by matipojo — APPROVED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Please pull from 3.5

File: modules/library/documents/page.php [hunk 17]
```diff
@@ -60,6 +62,10 @@ public static function get_plural_title() {
 		return __( 'Pages', 'elementor' );
 	}
 
+	public static function get_add_new_title() {
+		return __( 'Add New Library Page', 'elementor' );
```
Comment (rodolphebertozzo) [inline modules/library/documents/page.php:66]: esc

File: core/base/document.php [hunk 18]
```diff
@@ -167,6 +167,10 @@ public static function get_plural_title() {
 		return static::get_title();
 	}
 
+	public static function get_add_new_title() {
+		return sprintf( __( 'Add New %s', 'elementor' ), static::get_title() );
```
Comment (rodolphebertozzo) [inline core/base/document.php:171]: esc

**Review by rodolphebertozzo — CHANGES_REQUESTED** [review]

File: modules/library/documents/library-document.php [hunk 19]
```diff
@@ -44,6 +44,10 @@ public static function get_properties() {
 		$properties['admin_tab_group'] = 'library';
 		$properties['show_in_library'] = true;
 		$properties['register_type'] = true;
+		$properties['cpt'] = [ Source_Local::CPT ];
+
+		// TODO: Remove - Backwards compatibility, since 'Theme_Document' does not support 'static::get_type'.
+		$properties['show_in_finder'] = false;
```
Comment (matipojo) [inline modules/library/documents/library-document.php:50]: Is something will not work now?

**Review by matipojo — CHANGES_REQUESTED** [review]

File: modules/library/documents/library-document.php (hunk 19 above)
Comment (iNewLegend) [reply modules/library/documents/library-document.php:50]: Yes, `Theme_Document` calls eventually to 'get_site_editor_type_bc'. And produce error.

Comment (matipojo) [reply modules/library/documents/library-document.php:50]: Please create a PR the fix it and link it here.

**Review by matipojo — APPROVED** [review]

**Pull request merged**